### PR TITLE
Adds support for Github-style fenced math blocks in markdown editor and preview

### LINKED ISSUE #174731: Markdown math code block syntax should have LaTeX syntax highlighting inside
Type: <b>Bug</b>

When editing a Markdown file, VS Code will apply LaTeX syntax highlighting to math in double-dollar signs (math blocks), e.g.,
```
$$
\frac{1}{0} = \infty
$$
```
will color `frac` and `infty` and the numbers.

GitHub Markdown also supports a `math` code block syntax (the triple-tick syntax; see [reference](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/writing-mathematical-expressions#writing-expressions-as-blocks)), so the above could be written in GitHub-flavored Markdown as
```math
\frac{1}{0} = \infty
```
> GitHub will render the above as math, so here's what it looks like in Markdown:
> ```
> ```math
> \frac{1}{0} = \infty
> `'`
> ```

However, VS Code applies no syntax highlighting inside ```math blocks.

image

It'd be nice if it could apply the same highlighting as double-dollar signs.

---

VS Code version: Code 1.75.1 (Universal) (441438abd1ac652551dbe4d408dfcec8a499b8bf, 2023-02-08T21:34:59.000Z)
OS version: Darwin arm64 22.2.0
Modes:
Sandboxed: Yes

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Apple M1 (8 x 24)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: disabled_off<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>metal: disabled_off<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_renderer: enabled_on<br>video_decode: enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled<br>webgpu: disabled_off|
|Load (avg)|1, 2, 2|
|Memory (System)|16.00GB (0.28GB free)|
|Process Argv|--crash-reporter-id 842edd4e-1bf5-4a7a-9db0-8c5b7cec0f18|
|Screen Reader|no|
|VM|0%|
</details><details><summary>Extensions (19)</summary>

Extension|Author (truncated)|Version
---|---|---
prettier-vscode|esb|9.10.4
linter-gfortran|for|3.2.0
unicode-math-vscode|Gui|0.3.5
isort|ms-|2022.8.0
python|ms-|2023.2.0
vscode-pylance|ms-|2023.2.30
jupyter|ms-|2023.1.2010391206
jupyter-keymap|ms-|1.0.0
jupyter-renderers|ms-|1.0.14
vscode-jupyter-cell-tags|ms-|0.1.6
vscode-jupyter-slideshow|ms-|0.1.5
cmake-tools|ms-|1.13.45
cpptools|ms-|1.14.3
cpptools-extension-pack|ms-|1.3.0
vsliveshare|ms-|1.0.5831
rewrap|stk|1.16.3
cmake|twx|0.0.17
clang-format|xav|1.9.0
markdown-all-in-one|yzh|3.5.0

(1 theme extensions excluded)

</details><details>
<summary>A/B Experiments</summary>

```
vsliv368cf:30146710
vsreu685:30147344
python383cf:30185419
vspor879:30202332
vspor708:30202333
vspor363:30204092
vslsvsres303:30308271
pythonvspyl392:30443607
vserr242:30382549
pythontb:30283811
vsjup518:30340749
pythonptprofiler:30281270
vshan820:30294714
vstes263cf:30335440
vscoreces:30445986
pythondataviewer:30285071
vscod805cf:30301675
binariesv615:30325510
bridge0708:30335490
bridge0723:30353136
cmake_vspar411:30581797
vsaa593:30376534
pythonvs932:30410667
cppdebug:30492333
vsclangdf:30486550
c4g48928:30535728
dsvsc012:30540252
azure-dev_surveyone:30548225
vscccc:30610679
pyindex848:30662994
nodejswelcome1cf:30587006
2e4cg342:30602488
pyind779cf:30662993
89544117:30613380
pythonsymbol12:30657548
vsctsb:30662444

```

</details>

### PR DESCRIPTION
**Pull request by aaronchucarroll**
VSCode markdown editor now syntactically highlights fenced math blocks. Katex rendering now supports fenced math blocks. They must be of the form
```math
//math goes here
```

resolves #174731

### CODE REVIEW

**Comment by aaronchucarroll** [comment]
@microsoft-github-policy-service agree

**Review by mjbvz — CHANGES_REQUESTED** [review]
Thanks for looking into this. I'd like to keep all the code changes in the `markdown-math` extension as that's also where support for `$$` comes from

 You should be able to use a markdown-it plugin to do this

**Comment by aaronchucarroll** [comment]
With the latest commits the code changes are only located in markdown-math.

File: extensions/markdown-math/notebook/katex.ts [hunk 1]
```diff
@@ -2,7 +2,7 @@
  *  Copyright (c) Microsoft Corporation. All rights reserved.
  *  Licensed under the MIT License. See License.txt in the project root for license information.
  *--------------------------------------------------------------------------------------------*/
-import type * as markdownIt from 'markdown-it';
+import type MarkdownIt = require('markdown-it');
```
Comment (mjbvz) [inline extensions/markdown-math/notebook/katex.ts:5]: `require` generally shouldn't be used. Instead it's likely we should enable `esModuleInterop` in the `tsconfig`. I think the error message will tell you the exact option that's needed

File: extensions/markdown-math/src/extension.ts [hunk 2]
```diff
@@ -19,24 +21,34 @@ export function activate(context: vscode.ExtensionContext) {
 		return config.get<{ [key: string]: string }>('math.macros', {});
 	}
 
+	function preprocessMarkdown(markdown: string): string {
+		const fencedMathRegex = /```math\s+([\s\S]*?)\s+```/g;
```
Comment (mjbvz) [inline extensions/markdown-math/src/extension.ts:25]: I don't think a regular expression is the right approach. It's fragile and won't handle a lot of cases without extra work (a fetched code block can use more than 3 ``` ` ``` for example, or use `~~~`)

Instead can we override the fenced code block parser to handle the `math` language specially?

**Review by mjbvz — CHANGES_REQUESTED** [review]

File: extensions/markdown-math/src/extension.ts (hunk 2 above)
Comment (aaronchucarroll) [reply extensions/markdown-math/src/extension.ts:25]: Yea I can do this. Just to be clear we want to move the changes back to markdown-language-features, adding a type of fenced code block for math blocks?

**Comment by aaronchucarroll** [comment]
I added a custom parser for fenced math blocks based on the original markdown-it fence parser.

File: extensions/markdown-math/src/extension.ts [hunk 3]
```diff
@@ -2,41 +2,125 @@
...
+		const config = vscode.workspace.getConfiguration("markdown");
+		return config.get<{ [key: string]: string }>("math.macros", {});
 	}
 
-	vscode.workspace.onDidChangeConfiguration(e => {
-		if (e.affectsConfiguration(markdownMathSetting)) {
-			vscode.commands.executeCommand('markdown.api.reloadPlugins');
-		}
-	}, undefined, context.subscriptions);
+	vscode.workspace.onDidChangeConfiguration(
```
Comment (mjbvz) [inline extensions/markdown-math/src/extension.ts:97]: Please revert the extra formatting changes. They don't match this repo's style and make the PR more difficult to review

**Review by mjbvz — COMMENTED** [review]
Changes look ok besides the extra formatting changes

I'm curious though if this can be simplified even more. In the `vscode-markdown-it-katex` library, maybe we can override the rendering of fenced code blocks. Something like:

```
    const originalFenceRenderer = md.renderer.rules.fence;
    md.renderer.rules.fence = function (tokens, idx, options, env, self) {
        if (token is a math block, render it as a katex block) {
        return blockRenderer(tokens, idx);
        } else {
            return originalFenceRenderer?.call(this, tokens, idx, options, env, self);
        }
    };
```

Support for these should like be an new option, similar to the existing options in that library: https://github.com/mjbvz/vscode-markdown-it-katex/blob/e3497057288a70dd1dc14484cb1e258293b4e37c/src/index.ts#L462

File: extensions/markdown-math/src/extension.ts (hunk 3 above)
Comment (aaronchucarroll) [reply extensions/markdown-math/src/extension.ts:97]: Yea I must have changed my default formatter by mistake.

**Comment by aaronchucarroll** [comment]
This approach was much simpler. I made these changes in @vscode/markdown-it-katex, and added tests. In markdown-math I set enableFencedBlocks to be true by default but there may be a better way to do this. 

The PR to the katex library is [here](https://github.com/microsoft/vscode-markdown-it-katex/pull/17)

**Comment by aaronchucarroll** [comment]
I reverted all of the unnecessary changes from prior implementations. The main repo still needs the new syntax for syntax highlighting and enabling the fence option.

File: extensions/markdown-math/syntaxes/md-math-block.tmLanguage.json [hunk 4]
```diff
@@ -79,6 +82,42 @@
 					]
 				}
 			]
+		},
```
Comment (mjbvz) [inline extensions/markdown-math/syntaxes/md-math-block.tmLanguage.json:85]: For the grammar changes, I think we should create a separate injection grammar instead that uses this approach: https://github.com/steebn/vscode-fenced-code-block-grammar-injection-example/tree/master

It should reduce some of the grammar duplication and make sure we only render fenced code blocks in expected places

**Comment by aaronchucarroll** [comment]
Please see new changes (ignore accidental git mess)

**Review by mjbvz — APPROVED** [review]

**Review by DonJayamanne — APPROVED** [review]

**Pull request merged**